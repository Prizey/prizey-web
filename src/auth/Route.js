import React, { useLayoutEffect } from 'react'
import { connect } from 'react-redux'
import { navigate } from '@reach/router'

export const getState = state => ({ state })

export default connect(
  getState,
  null,
)(
  ({
    Component,
    authorize,
    unauthorized,
    currentUser,
    state,
    location,
    ...props
  }) => {
    useLayoutEffect(() => {
      if (!currentUser) {
        navigate(`/sign-in?next=${location.pathname}`)
      } else if (currentUser.blocked) {
        navigate('/blocked')
      }

      if (authorize) {
        authorize(state, currentUser) || navigate(unauthorized)
      }
    }, [authorize, currentUser, location, state, unauthorized])

    useLayoutEffect(() => {
      window.scrollTo(0, 0)
    }, [location])

    return currentUser ? (
      <Component {...props} currentUser={currentUser} location={location} />
    ) : null
  },
)
