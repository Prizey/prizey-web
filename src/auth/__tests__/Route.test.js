import React from 'react'
import renderer from 'react-test-renderer'
import { navigate } from '@reach/router'

import Route, { getState } from '../Route'

jest.mock('react-redux', () => ({
  connect: () => Component => props => <Component {...props} state={{}} />,
}))

jest.mock('@reach/router', () => ({
  navigate: jest.fn(),
}))
window.scrollTo = jest.fn()

const Component = () => <div>Foo bar foo</div>
const location = { pathname: 'fooURL' }

it('test the get of state', () => {
  const state = { basket: {} }
  expect(getState(state)).toEqual({
    state,
  })
})

describe('without authenticated user', () => {
  it('redirect user to signIn route', () => {
    renderer.create(<Route Component={Component} location={location} />)

    expect(navigate).toHaveBeenCalledWith(`/sign-in?next=${location.pathname}`)
  })
})

describe('whith currentUser', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <Route
          Component={Component}
          location={location}
          currentUser={{ id: 1, name: 'foo' }}
        />,
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('redirect to blocked when the currentUser is blocked', () => {
    renderer.create(
      <Route
        Component={Component}
        location={location}
        currentUser={{ blocked: true, id: 1, name: 'foo' }}
      />,
    )
    expect(navigate).toHaveBeenCalledWith('/blocked')
  })

  it('the page should be scrolled to the top', () => {
    renderer.create(
      <Route
        Component={Component}
        location={location}
        currentUser={{ id: 1, name: 'foo' }}
      />,
    )
    expect(window.scrollTo).toHaveBeenCalledWith(0, 0)
  })
})
