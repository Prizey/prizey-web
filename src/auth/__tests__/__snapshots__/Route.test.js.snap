// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`whith currentUser renders correctly 1`] = `
<div>
  Foo bar foo
</div>
`;
